plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/actions/SetDirtyStateAction.java

```diff
@@ -20,18 +20,24 @@
    public static final String ID = "setDirtyState";
 
    private boolean isDirty;
+   private Action causedBy;
 
    public SetDirtyStateAction() {
       super(ID);
    }
 
-   public SetDirtyStateAction(final boolean isDirty) {
+   public SetDirtyStateAction(final boolean isDirty, final Action causedBy) {
```

Comment:
Do we have to do this a mandatory parameter? Afaik, VSCode just wants to know that for undo actions, right? I'm wondering whether this justifies making this a mandatory parameter for all other invocations caused by anything else than `UndoAction`. Also do you think it would be simpler to just have a string argument: `reason = 'operation' | 'undo' | 'save'`? I don't have a hard opinion on that, but enforcing to provide an actual action seems like a strong requirement for this rather specific use case of VSCode.

Minor knitpick, I think as a name I'd prefer `reason` rather than `causedBy`.

Reply:
+1 for `reason` ;-)

Reply:
+1 for 'reason' and making it optional. In any case it should be possible to determine wehter the dirty state change was triggered by an actual new edtit/operation, an undo/redo action or a save action. Now thinking about it, we also have a another case: the server could also update the model without an operation invocation (e.g. by receiving an update from the model server).
So maybe we should type reason as follows: 'operation'|'undo'|'redo'|'save'|'externalUpdate'.
What do you think?

Reply:
:+1: Yes, sounds good! Maybe just `external` is good enough, e.g. also if the file got updated on the file system or whatever other external reason. Thanks!

Reply:
Maybe it should be just String then and adopters can provide any reason they want and we just have a list of known String reasons but no strict typing?

Reply:
Fine with me as well. We could then just use the action.kind as default reason string as well